pkg/ceph/osd/filesystem.go

```diff
@@ -0,0 +1,244 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/rook/rook/pkg/ceph/mon"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/util"
+)
+
+const (
+	maxFileBackupSize         = 1 * 1024 * 1024 // 1 MB
+	osdFSStoreNameFmt         = "rook-ceph-osd-%d-fs-backup"
+	bluestoreBlockSymlinkName = "block"
+	bluestoreDBSymlinkName    = "block.db"
+	bluestoreWalSymlinkName   = "block.wal"
+)
+
+// creates/initalizes the OSD filesystem and journal via a child process
+func createOSDFileSystem(context *clusterd.Context, clusterName string, config *osdConfig) error {
+	logger.Infof("Initializing OSD %d file system at %s...", config.id, config.rootPath)
+
+	// get the current monmap, it will be needed for creating the OSD file system
+	monMap, err := getMonMap(context, clusterName)
+	if err != nil {
+		return fmt.Errorf("failed to get mon map: %+v", err)
+	}
+
+	// the current monmap is needed to create the OSD, save it to a temp location so it is accessible
+	monMapTmpPath := getOSDTempMonMapPath(config.rootPath)
+	monMapTmpDir := filepath.Dir(monMapTmpPath)
+	if err := os.MkdirAll(monMapTmpDir, 0744); err != nil {
+		return fmt.Errorf("failed to create monmap tmp file directory at %s: %+v", monMapTmpDir, err)
+	}
+	if err := ioutil.WriteFile(monMapTmpPath, monMap, 0644); err != nil {
+		return fmt.Errorf("failed to write mon map to tmp file %s, %+v", monMapTmpPath, err)
+	}
+
+	options := []string{
+		"--mkfs",
+		fmt.Sprintf("--id=%d", config.id),
+		fmt.Sprintf("--cluster=%s", clusterName),
+		fmt.Sprintf("--conf=%s", mon.GetConfFilePath(config.rootPath, clusterName)),
+		fmt.Sprintf("--osd-data=%s", config.rootPath),
+		fmt.Sprintf("--osd-uuid=%s", config.uuid.String()),
+		fmt.Sprintf("--monmap=%s", monMapTmpPath),
+		fmt.Sprintf("--keyring=%s", getOSDKeyringPath(config.rootPath)),
+	}
+
+	if !isBluestore(config) {
+		options = append(options, fmt.Sprintf("--osd-journal=%s", getOSDJournalPath(config.rootPath)))
+	}
+
+	// create the OSD file system and journal
+	err = context.ProcMan.Run(
+		fmt.Sprintf("mkfs-osd%d", config.id),
+		"ceph-osd",
+		options...)
+
+	if err != nil {
+		return fmt.Errorf("failed osd mkfs for OSD ID %d, UUID %s, dataDir %s: %+v",
+			config.id, config.uuid.String(), config.rootPath, err)
+	}
+
+	// now that the OSD filesystem has been created, back it up so it can be restored/repaired
+	// later on if needed.
+	if err := backupOSDFileSystem(config, clusterName); err != nil {
+		return fmt.Errorf("failed to backup OSD filesystem: %+v", err)
+	}
+
+	// update the scheme to indicate the OSD's filesystem has been created
```

Comment:
should we mark the filesystem created before the backup? Just in case the backup fails.

Reply:
the semantics of `isOSDFilesystemCreated` actually mean "is the file system created AND backed up so that if we lose it we can restore it again?".  if we were to mark it as created (completed) before backing it up and then the back up fails, we'll think that we can restore it from a backup that we don't actually have.

perhaps i should change the naming to "completed" or "ready" instead of "created"?  would that work?

Reply:
I will add a comment to clarify